src/main/java/uk/ac/sanger/sccp/stan/service/SampleProcessingServiceImp.java

```diff
@@ -0,0 +1,100 @@
+import uk.ac.sanger.sccp.stan.request.OperationResult;
+
+import java.util.*;
+
+import static java.util.Objects.requireNonNull;
+
+@Service
+public class SampleProcessingServiceImp implements SampleProcessingService {
+    private final SampleRepo sampleRepo;
+    private final TissueRepo tissueRepo;
+    private final OperationTypeRepo opTypeRepo;
+    private final LabwareRepo labwareRepo;
+    private final OperationService opService;
+    private final Validator<String> externalNameValidator;
+
+    @Autowired
+    public SampleProcessingServiceImp(SampleRepo sampleRepo,
+                                      TissueRepo tissueRepo,
+                                      OperationTypeRepo opTypeRepo,
+                                      LabwareRepo labwareRepo,
+                                      OperationService opService,
+                                      @Qualifier("externalNameValidator") Validator<String> externalNameValidator) {
+        this.sampleRepo = sampleRepo;
+        this.tissueRepo = tissueRepo;
+        this.opTypeRepo = opTypeRepo;
+        this.labwareRepo = labwareRepo;
+        this.opService = opService;
+        this.externalNameValidator = externalNameValidator;
+    }
+
+    @Override
+    public OperationResult addExternalID(User user, AddExternalIDRequest request) {
+        requireNonNull(user, "User is null.");
+        requireNonNull(request, "Request is null.");
+        Collection<String> problems = new LinkedHashSet<>();
+
+        Labware lw = labwareRepo.getByBarcode(request.getLabwareBarcode());
+        String externalName = request.getExternalName();
+        List<Sample> samples = new ArrayList<>();
+        lw.getSlots().forEach((slot) -> {
+            if (slot.getSamples() != null) {
+                samples.addAll(slot.getSamples());
+            }
+        });
+
+        validateSamples(problems, samples);
+        validateExternalName(problems, externalName);
+
+        if (!problems.isEmpty()) {
+            throw new ValidationException("The request could not be validated.", problems);
+        }
+
+        Tissue tissue = samples.get(0).getTissue();
+        tissue.setExternalName(externalName);
+        OperationType opType = opTypeRepo.getByName("Add external ID");
+        Operation op = opService.createOperationInPlace(opType, user, lw, null, null);
+        return new OperationResult(List.of(op), List.of(lw));
+    }
+
+    public void validateSamples(Collection<String> problems, List<Sample> samples) {
+        if (samples.isEmpty()) {
+            problems.add("Could not find a sample associated with this labware");
+            return;
+        }
+        if (samples.size() > 1) {
+            problems.add("There are too many samples associated with this labware");
+            return;
+        }
+        Tissue tissue = samples.get(0).getTissue();
+        if (!tissue.getExternalName().isEmpty() || tissue.getExternalName() == null) {
```

Comment:
The logic is faulty here: you are checking if the name is NOT empty, or if it IS null. I think you mean
```
(tissue.getExternalName()!=null && !tissue.getExternalName().isEmpty())
```